telegram/ext/_dispatcher.py

```diff
@@ -282,6 +282,27 @@ def _set_singleton(cls, val: Optional['Dispatcher']) -> None:
         cls.logger.debug('Setting singleton dispatcher as %s', val)
         cls.__singleton = weakref.ref(val) if val else None
 
+    @property
+    def chat_data(self) -> Mapping[int, CD]:
+        """
+        :obj:`types.MappingProxyType`: A dictionary handlers can use to store data for the chat.
+
+        .. versionchanged:: 14.0
+           :attr:`chat_data` is now read-only
+
+        """
+        return MappingProxyType(self._chat_data)
```

Comment:
hm so no property then

Reply:
Turns out it might be necessary. Consider this snippet:

``` python
class B:
    def __init__(self):
        self._chat_data = defaultdict(list)
        self.chat_data = MappingProxyType(self._chat_data)


b = B()
b._chat_data = defaultdict(dict)
b._chat_data.update({1: {}})

print(b._chat_data, b.chat_data)
# prints: defaultdict(<class 'dict'>, {1: {}}) defaultdict(<class 'list'>, {})
```
This is why the tests are failing for me, the `self.chat_data` mapping does not update when `self._chat_data` is reassigned (if it simply updated, it works fine)

Anyway, the performance hit of recreating the object every time is small. Unless you have another idea to make it work?

Reply:
ah okay. that's to be expected, I guess …  IISC the only place where we override `_chat_data` is while loading the persistence data and I see several possible solutions here:

1. just rebuild `self.chat_data = MPT(self._chat_data)` after overriding at this one place
2. make `self._chat_data` a property with a setter that does the rebuilding
3. think about if we really need to override `chat_data` on loading the persistence. The more I look at it, the stranger it seems to me that we do that instead of `update(get_chat_data())`. Switching to `update` should even be non-breaking, I guess (even though we might be able to able to simplify some internals of PP/DictPersistence and adjust the expected return types of `get_*_data` - can be done in a separate PR, though)

what do you think?

Reply:
> 1. just rebuild `self.chat_data = MPT(self._chat_data)` after overriding at this one place

might work for now, but might be annoying to keep doing that in the long term - specially in tests

> 2. make `self._chat_data` a property with a setter that does the rebuilding

yes, sounds like it could work well, even when users modify/reassign `dp._chat_data`.

> 3. think about if we really need to override `chat_data` on loading the persistence. The more I look at it, the stranger it seems to me that we do that instead of `update(get_chat_data())`. Switching to `update` should even be non-breaking, I guess (even though we might be able to able to simplify some internals of PP/DictPersistence and adjust the expected return types of `get_*_data` - can be done in a separate PR, though)

yep, makes sense to use `defaultdict` in one place and just update it than do it two other places and reassign it.

Reply:
ok I'm stupid... a property for `dp._chat_data` wouldn't work either, it just leads to the initial problem. I ended up adding methods `_set_chat/user_data` which would also update `dp.chat/user_data`